runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectGetWorkStream.java

```diff
@@ -250,6 +257,22 @@ public GetWorkBudget remainingBudget() {
         .apply(currentInflightBudget);
   }
 
+  @Override
+  public synchronized void close() {
+    super.close();
+    if (!rpcChannel.isShutdown()) {
```

Comment:
it seems like this could shutdown a channel that is still in the cache?
It seems like the cache should be responsible for shutting down channels that it is sure that are not being used anymore.

If we need to track who is using the endpoint, do we need the cache? It seems like instead we could have the StreamingEngineClient cache the channels internally and shut them down when the worker endpoints change.

Reply:
yea this is what was i was initially thinking but opted to use the cache and handle the closing of the channels in `StreamingEngineClient`.

I think the channelCache encapsulates the logic well and makes it more testable?